Extended performance tips about abstract types (#34946)
JuliaLang/julia: doc/src/manual/performance-tips.md

the type is fully-specified in the first case, the compiler doesn't need to generate any code
to resolve the type at run-time. This results in shorter and faster code.

> One should also keep in mind that not-fully-parameterized types behave like abstract types. For example, even though `Array` is not an abstract type, it is not a concrete type either when its parameters are not bound:

Reviewer: ```suggestion
One should also keep in mind that not-fully-parameterized types behave like abstract types. For example, even though a fully specified `Array{T,n}` is concrete, `Array` itself with no parameters given is not concrete:
```

```jldoctest myambig3

> julia> isconcretetype(Array), isabstracttype(Array), isstructtype(Array), isconcretetype(Array{Int}), isconcretetype(Array{Int,1})
> (false, false, true, false, true)

Reviewer 2: ```suggestion
julia> !isconcretetype(Array), !isabstracttype(Array), isstructtype(Array), !isconcretetype(Array{Int}), isconcretetype(Array{Int,1})
(true, true, true, true, true)
```

Author: I'm not sure about this, it seems a bit harder to decode for the reader imho.

Reviewer 2: My think was that it would be helpful that all of these are testing true, so the reader doesn't need to take the negation in their head

Reviewer 3: In this case, to emphasize the all-true-tests, perhaps:
```julia
julia> !isconcretetype(Array) && !isabstracttype(Array) && isstructtype(Array) && !isconcretetype(Array{Int}) && isconcretetype(Array{Int,1})
true
```
?

Author: It's just that I find `false` more readable than a `!` in front of a function.

the type is fully-specified in the first case, the compiler doesn't need to generate any code
to resolve the type at run-time. This results in shorter and faster code.

One should also avoid fields with unparameterized parametric types:

```jldoctest myambig3
julia> mutable struct MyParametricType{T}
           a::T
       end
       
julia> mutable struct MySecondType{T <: MyParametricType}
           mpt::T
       end
       
julia> mutable struct MyThirdType{T}
           mpt::MyParametricType{T}
       end
 ```
`MySecondType` and `MyThirdType` are both better choices than 

```jldoctest myambig3
julia> mutable struct MyAlsoAmbiguousType
           mpt::MyParametricType
       end
```

> Even though `MyParametricType` is not an abstract type, when it is not parameterized, it is not a concrete type either:

Reviewer 4: > `when it is not parameterized`

I read four times until I realized this belongs to `it is not a concrete type either` 😂  Perhaps you can adjust it a bit for nonnative English speakers.

Author: I guess this is because I'm not native either. I changed this to something clearer.
